br/pkg/lightning/backend/local/duplicate.go

```diff
@@ -60,65 +63,184 @@ type DuplicateRequest struct {
+		return bytes.Compare(indexHandles.rawHandles[i], rawHandle) >= 0
+	})
+}
+
+// physicalTableIDs returns all physical table IDs associated with the tableInfo.
+// A partitioned table can have multiple physical table IDs.
+func physicalTableIDs(tableInfo *model.TableInfo) []int64 {
+	if tableInfo.Partition != nil {
+		defs := tableInfo.Partition.Definitions
+		tids := make([]int64, 1, len(defs)+1)
+		tids[0] = tableInfo.ID
+		for _, def := range defs {
+			tids = append(tids, def.ID)
+		}
+		return tids
+	}
+	return []int64{tableInfo.ID}
+}
+
+// NewDuplicateManager creates a new *DuplicateManager.
+//
+// This object provides methods to collect and decode duplicated KV pairs into row data. The results
+// are stored into the errorMgr.
 func NewDuplicateManager(
-	db *pebble.DB,
+	errorMgr *errormanager.ErrorManager,
 	splitCli restore.SplitClient,
 	ts uint64,
 	tls *common.TLS,
 	regionConcurrency int) (*DuplicateManager, error) {
 	return &DuplicateManager{
-		db:                db,
+		errorMgr:          errorMgr,
 		tls:               tls,
 		regionConcurrency: regionConcurrency,
 		splitCli:          splitCli,
 		keyAdapter:        duplicateKeyAdapter{},
 		ts:                ts,
 		connPool:          common.NewGRPCConns(),
+		// TODO: not sure what is the correct concurrency value.
+		remoteWorkerPool: utils.NewWorkerPool(uint(regionConcurrency), "duplicates"),
 	}, nil
 }
 
-func (manager *DuplicateManager) CollectDuplicateRowsFromTiKV(ctx context.Context, tbl table.Table) error {
-	log.L().Info("Begin collect duplicate data from remote TiKV")
+// CollectDuplicateRowsFromTiKV collects duplicated rows already imported into TiKV.
+//
+// Collection result are saved into the ErrorManager.
+func (manager *DuplicateManager) CollectDuplicateRowsFromTiKV(ctx context.Context, tbl table.Table) (hasDupe bool, err error) {
+	logTask := log.L().Begin(zapcore.InfoLevel, "collect duplicate data from remote TiKV")
+	defer func() {
+		logTask.End(zapcore.InfoLevel, err)
+	}()
+
 	reqs, err := buildDuplicateRequests(tbl.Meta())
 	if err != nil {
-		return err
+		return false, err
 	}
 
+	// TODO: reuse the *kv.SessionOptions from NewEncoder for picking the correct time zone.
 	decoder, err := kv.NewTableKVDecoder(tbl, &kv.SessionOptions{
 		SQLMode: mysql.ModeStrictAllTables,
 	})
 	if err != nil {
-		return err
+		return false, err
 	}
 	g, rpcctx := errgroup.WithContext(ctx)
+	atomicHasDupe := atomic.NewBool(false)
 	for _, r := range reqs {
 		req := r
-		g.Go(func() error {
-			err := manager.sendRequestToTiKV(rpcctx, decoder, req)
+		manager.remoteWorkerPool.ApplyOnErrorGroup(g, func() error {
+			err := manager.sendRequestToTiKV(rpcctx, decoder, req, atomicHasDupe)
 			if err != nil {
 				log.L().Error("error occur when collect duplicate data from TiKV", zap.Error(err))
 			}
 			return err
 		})
 	}
-	err = g.Wait()
-	log.L().Info("End collect duplicate data from remote TiKV")
-	return err
+	return atomicHasDupe.Load(), errors.Trace(g.Wait())
```

Comment:
According to https://golang.org/ref/spec#Order_of_evaluation, `atomicHasDupe.Load()` will be evaluated before `g.Wait()`, `atomicHasDupe` may not record if there are duplicates.